### PR TITLE
test: Code Coverage: Add Edit Mileage: #8 Part 1

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 7765554</samp>

This pull request enhances the testing and functionality of the `add-edit-mileage` component and the mileage expense feature. It adds and updates test cases using new mock data and custom fields, and it adds optional chaining operators to prevent errors when accessing form values. It also adds new mock data for cost centers, org settings, and unflattened transactions to simulate different scenarios of mileage expenses.

### <samp>🤖 Generated by Copilot at 7765554</samp>

> _`Add-edit-mileage` is the path of doom_
> _Mocking the APIs with cost centers and custom fields_
> _Unflattened transactions, complex and dark_
> _Optional chaining operators, a spark of hope in the code_

### <samp>🤖 Generated by Copilot at 7765554</samp>

*  Add mock data for cost centers, org settings, and unflattened transactions ([link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-7de0329a6c26862e485c2274615a2f69250e4144bdf3abd27cc61d704384ec91R285-R339), [link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-1cac010b51c1a7c2fbdf9bf224a7e267779365abe08b0f16df1ef46e41483951R1309-R1315), [link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-21ae928ce640ad62cfada5fee7bf0bbc490300ab04a25f8309a9878e21b2f1cfR2883-R3243))
*  Import mock data and update test cases for add-edit-mileage component ([link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-b97456a1337050be58f0238bfa7892f70e10c42a7d58754c5d1528a1a0952d7fL55-R65), [link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-b97456a1337050be58f0238bfa7892f70e10c42a7d58754c5d1528a1a0952d7fL65-R87), [link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-b97456a1337050be58f0238bfa7892f70e10c42a7d58754c5d1528a1a0952d7fR541-R698))
*  Add optional chaining operators to prevent errors when accessing form values in `add-edit-mileage.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-04c216da6200af711aa0f53bcd7e253086762b5d123b0729fac3b1862ec5fc3eL2058-R2062), [link](https://github.com/fylein/fyle-mobile-app/pull/2333/files?diff=unified&w=0#diff-04c216da6200af711aa0f53bcd7e253086762b5d123b0729fac3b1862ec5fc3eL2078-R2086))

## Clickup
https://app.clickup.com/t/85ztjmtfq

## Code Coverage
`74.88% Statements 626/836`
`71.96% Branches 534/742`
`77.05% Functions 225/292`
`75.37% Lines 600/796`

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts [hunk 1]
```diff
@@ -521,5 +538,165 @@ export function TestCases4(getTestBed) {
...
+          },
+        ]);
+
+        component.getReports().subscribe((res) => {
+          expect(res).toEqual(expectedErpt[0]);
+          done();
+        });
+      });
+
+      it("should return the first report if there's only one option in DRAFT state", (done) => {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts:559]: use single quotes

File: src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts [hunk 2]
```diff
@@ -521,5 +538,165 @@ export function TestCases4(getTestBed) {
...
+      fixture.detectChanges();
+
+      const result = component.getCostCenterDependentFields();
+      expect(result).toEqual([]);
+    });
+
+    describe('generateEtxnFromFg():', () => {
+      it('should generate an expense from form', (done) => {
+        dateService.getUTCDate.and.returnValue(new Date('2023-02-13T01:00:00.000Z'));
+        spyOn(component, 'getFormValues').and.returnValue({
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts:670]: add this to a separate file

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts [hunk 3]
```diff
@@ -2055,11 +2055,11 @@ export class AddEditMileagePage implements OnInit {
           tx: {
             ...etxn.tx,
             mileage_vehicle_type: formValue.mileage_rate_name?.vehicle_type,
-            mileage_is_round_trip: formValue.route.roundTrip,
+            mileage_is_round_trip: formValue?.route.roundTrip,
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:2058]: aren't we assigning the `formValue` above? 
according to this change, 
for `mileage_vehicle_type: formValue.mileage_rate_name?.vehicle_type,` we know `formValue` is present, 
but the immediate next line we are checking again if it's present or not. We shouldn't be doing this no?

**Review by Dimple16 — CHANGES_REQUESTED** [review]
added comments

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts (hunk 3 above)
Comment (jayfyle) [reply src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts]: Removed the extra optional chaining

**Review by Dimple16 — APPROVED** [review]

File: src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts [hunk 4]
```diff
@@ -516,5 +530,152 @@ export function TestCases4(getTestBed) {
...
+    });
+
+    describe('getEditRates():', () => {
+      it('should return mileage rate from expense', fakeAsync(() => {
+        component.etxn$ = of(newExpenseMileageData2);
+        component.mileageRates$ = of(unfilteredMileageRatesData);
+        fixture.detectChanges();
+
+        component.getEditRates().subscribe((res) => {
+          expect(res).toEqual(10);
```
Comment (OmkarJ13) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts:604]: Do we not need `done()` here?

File: src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts [hunk 5]
```diff
@@ -516,5 +530,152 @@ export function TestCases4(getTestBed) {
...
+        project_dependent_fields: [],
+      });
+      fixture.detectChanges();
+
+      const result = component.getProjectDependentFields();
+      expect(result).toEqual([]);
+    });
+
+    it('getCostCenterDependentFields(): get cost center dependent fields from the form', () => {
+      component.fg.patchValue({
```
Comment (OmkarJ13) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-4.spec.ts:650]: Why are we defining `project_dependent_fields`, when we are testing the cost center ones? Am I missing something?

**Review by OmkarJ13 — APPROVED** [review]

**Review by Chethan-Fyle — APPROVED** [review]

**Pull request merged**